### ISSUE TITLE
Nodebox compatibility - does scale() affect the origin of shapes ?

### ISSUE THREAD

**Issue by stuaxo** [shoebot/shoebot#436]
I noticed that scale() doesn't affect placement, e.g. draw a load of rectangles on a grid + scale, and the top-left corner of each rectangle appears in the same place.

Need to check this in Nodebox, as it doesn't seem like this can be right ?

**Comment by karstenw** [comment]
I tried it in NB with this

```
size(800,800)
align(CORNER)
# transform( (0,0) )
scale( 1 )

steps = 20
stepsize = WIDTH / steps / 2
nofill()
stroke(0)
strokewidth( 1 )
for r in range( steps ):
    rect( r * stepsize , r * stepsize, 10, 10 )
```

With the script as shown, the rect centers stay the same and the sizes vary.
Adding a "transform( (0,0) )" gives the expected behaviour of scaled coordinates.

Don't ask me why. Transforms et al have always puzzled me.
 
Scale 1
![rect-scale1](https://user-images.githubusercontent.com/652583/217511025-8e9dffa0-4dee-4ceb-a349-cde5479a0469.png)

Scale 2
![rect-scale2](https://user-images.githubusercontent.com/652583/217511016-29ceaabe-09c3-4678-90b6-529e26780837.png)

Scale2 + transform
![rect-scale2+transform](https://user-images.githubusercontent.com/652583/217511586-9a00e8aa-d5f9-4ee2-ab62-5a1ef507eb1f.png)

**Comment by karstenw** [comment]
Sorry. I confused transform & translate and align. My bad.

using `transform(CORNER)` gives the expected result; `transform(CENTER)` seems to be default.

**Comment by stuaxo** [comment]
I ran this in shoebot (after adding a draw(): since it needs shoebot needs a little fixing without it) - but was left confused enough I'll need to go to pen and paper to work out what I expect to happen :)

**Comment by karstenw** [comment]
I played around and found out:
- to apply translate & scale as expected: draw at 0,0 and let translate & scale handle it
- - keep drawing commands relative
- use absolute coordinates only without transform

This snippet works as expected in NB:
```
scale( 1.0 )
for r in range( steps ):
    rect( 0,0, 10,10 )
    translate( stepsize, stepsize )
```

**Comment by stuaxo** [comment]
I was just trying to draw a 10x10 grid that filled the window, weirdly difficult.

```python
from random import seed, random

def setup():
    var("gridsize", NUMBER, default=10, min=1, max=32)

def draw():
    nofill()
    colormode(HSB)
    
    scale(WIDTH / gridsize)
    translate(-.5, .5)
    
    for y in range(0, int(gridsize)+1):
        seed(y)
        push()
        for x in range(0, int(gridsize)+1):
            r = random()
            stroke(r, 1, 1)
            fill(r, 1, 1, .5)
            translate(1, 0)
            rect( 0,0, 1, 1 )
        pop()
        translate(0, 1)
```

Reminds me of a thought experiment I had about a non nodebox dialect where the default for every API would be centred shapes of size 1.0 at 0.0.

Maybe the size of the canvas would default to 1.0, with 0.0, 0.0 in the centre.

I guess that's a bit closer to cairo or opengl work natively; it feels like you could have quite a minimal drawing API.

**Comment by stuaxo** [comment]
Back on topic... 

I guess I should probably check out the nodebox grid API and see how good that is for filling a grid with items + maybe look at more examples for that.

**Comment by karstenw** [comment]
> Reminds me of a thought experiment I had about a non nodebox dialect where the default for every API would be centred shapes of size 1.0 at 0.0.

You should definitely read Peter Hendersons "Functional Geometry" Papers... like [here](https://shashi.biz/ijulia-notebooks/funcgeo/)

**Comment by karstenw** [comment]
I am having real fun with this. My heavily edited & experimented NB version.

```
import random as rnd
seed = rnd.seed

size( 400, 400 )

def handler(val, name):
    global gridsize
    if name == "gridsize":
        gridsize = int(round(val))
        # print( name, gridsize )
        draw()

var("gridsize", NUMBER, default=10, min=1, max=32, handler=handler)


def draw():
    nofill()
    colormode(HSB)
    # scale(WIDTH / gridsize)
    # translate(-.5, .5)

    #print("gridsize", gridsize)
    #print("scale:", WIDTH / gridsize )

    # old
    linewidth = 1.0 / gridsize
    boxwidth = 1.0 - linewidth

    # new
    cellsize = int( WIDTH / gridsize )
    linewidth = cellsize / 10.0
    boxwidth = cellsize - linewidth
    strokewidth( linewidth )
    
    for yi in range( gridsize ):
        y = yi * cellsize
        seed(y)
        #push()
        for xi in range( gridsize ):
            x = xi * cellsize
            r = rnd.random()
            stroke(r, 1, 1)
            fill(r, 1, 1, .5)
            #translate(1, 0)
            rect( x,y, boxwidth, boxwidth )
        #pop()
        #translate(0, 1)
    # print(_ctx._transform.matrix)
draw()

```
![colorgrid2](https://user-images.githubusercontent.com/652583/217800559-e7f5531e-7d2d-44f3-896c-b418664f5079.png)

Seeing the image here It's still missing an initial `translate( linewidth/2, linewidth/2 )`

**Comment by karstenw** [comment]
My final entry in the "Funny colored square" competition ;-) NodeBox style...


![colorgrid3](https://user-images.githubusercontent.com/652583/217807499-25cd3319-516e-4b98-9eb1-2fbd7d06cd53.png)

```python
import random as rnd
seed = rnd.seed

canvassize = 400
size( canvassize, canvassize )

background(1)

def handler(val, name):
    global gridsize
    gridsize = int(round(val))
    draw()

var("gridsize", NUMBER, default=10, min=1, max=32, handler=handler)

def draw():
    cellsize = int( WIDTH / gridsize )
    linewidth = cellsize / 10.0
    linemiddle = linewidth / 2
    boxwidth = cellsize - linewidth
    nofill()
    colormode(HSB)
    strokewidth( linewidth )
    translate( linemiddle, linemiddle )
    for row in range( gridsize ):
        y = row * cellsize
        seed( row )
        for column in range( gridsize ):
            x = column * cellsize
            hue = random()
            stroke(hue, 1, 1)
            fill(hue, 1, 1, 0.5)
            rect( x,y, boxwidth, boxwidth )
draw()

```

**Comment by stuaxo** [comment]
I will definitely have a read, that paper looks good.

What I was playing with last night, was - keeping it centered, + making the colours not jump around when you resize the grid.

The trick seems to be to make the seed stable in the center of the grid, then change it going outwards, the code I ended up with wasn't that nice, so I probably need to poke it some more.

I'll try these with your updates, taking the line width into account is a good - less lazy than my version.

**Comment by karstenw** [comment]
> What I was playing with last night, was - keeping it centered, + making the colours not jump around when you resize the grid.

I was thinking about that too. My idea would have been a pre-computed color list of gridsizemax * gridsizemax entries or using a dict with (xi,yi) entries. Havent followed that because the seed() in the y-loop keeps it nice enough. NodeBox reuses the last seed with VARs.


![colorgrid3-loop-00001](https://user-images.githubusercontent.com/652583/217834337-d68747c2-b738-4093-b19d-f7f12b89e43b.gif)

**Comment by stuaxo** [comment]
~Running~ Tying to run this on shoebot and finding holes in the `var` support :D

**Comment by karstenw** [comment]
Have you found out if scale() works as expected?

I am still playing with this stuff...

https://user-images.githubusercontent.com/652583/218271987-ed249de1-a910-4869-b89f-ac5c83756ef2.mp4


```python
from random import seed

canvassize = 400
size( canvassize, canvassize )
background(1)

def trisubdivide( cell ):
    """Divide a square cell (left,top,width) into 9 subcells"""
    left, top, width = cell
    zeronada = 0.0
    onethird = width / 3.0
    twothird = onethird * 2
    return (
        # center
        (left + onethird, top + onethird, onethird),
        # north
        (left + onethird, top + zeronada, onethird),
        # north-east
        (left + twothird, top + zeronada, onethird),
        # east
        (left + twothird, top + onethird, onethird),
        # south-east
        (left + twothird, top + twothird, onethird),
        # south
        (left + onethird, top + twothird, onethird),
        # south-west
        (left + zeronada, top + twothird, onethird),
        # west
        (left + zeronada, top + onethird, onethird),
        # north-west
        (left + zeronada, top + zeronada, onethird) )

def walkdown( initialcells, level=3, keepoldgenerations=True ):
    result = []
    if keepoldgenerations:
        result = initialcells[:]
    cells = initialcells
    for i in range( level ):
        levelcells = []
        for cell in cells:
            levelcells.extend( trisubdivide( cell ) )
        cells = levelcells[:]
        if keepoldgenerations:
            result.extend( cells )
        else:
            result = cells
    return result

def draw():
    inset = 10
    cellsize = WIDTH - (2 * inset)
    
    # canvas center
    mx, my = int(WIDTH/2), int(HEIGHT/2)

    initialcells = [ (inset,inset, cellsize) ]

    cells = walkdown( initialcells, 2, 0 )
    print("cells:", len(cells))

    colormode(HSB)
    transform( CORNER )

    for cell in cells:
        print( cell )
        hue = random()
        stroke(hue, 1, 1)
        fill(hue, 1, 1, 0.5)
        
        left,top,width = cell
        
        push()

        linesize = 0.02
        cellwidth = 1.0 - linesize
        strokewidth( linesize )

        translate(left, top)
        scale( width )

        rect( 0,0, cellwidth,cellwidth )
        # rect( 0,0, 1,1 )
        pop()
draw()

```

**Comment by stuaxo** [comment]
Ooh, that's nice.

I have discovered that I consistently get scale wrong when building scripts to play with, but I need to write some test bots to see if they are doing what Nodebox does.

I feel like setting scale in the global scope or in setup should probably set the initial scale, and not sure shoebot is preserving this.

I got a bit distracted trying to work out what an API that sets the scale so that N units fits across whatever width there already is.

Either another kw Param to scale, or something new like `scaletofit` or something better.

**Comment by stuaxo** [comment]
Ooh, that's nice.

I have discovered that I consistently get scale wrong when building scripts to play with, but I need to write some test bots to see if they are doing what Nodebox does.

I feel like setting scale in the global scope or in setup should probably set the initial scale, and not sure shoebot is preserving this.

I got a bit distracted trying to work out what an API that sets the scale so that NxM units fits across whatever width/height there already is.

Either another kw Param to scale, or something new like `scaletofit` or something better named.

**Comment by stuaxo** [comment]
Some feedback would probably help, I guess this isn't really a bug.

Started looking at a little drawgrid routing you can import (and at the same time was reminded about how ugly `_ctx` looks, but I'm not sure there is a good solution for that (especially in the nodebox1/shoebot style) :)

```python
# grid.py
def drawgrid():
    gridsize = 100
    _ctx.stroke(0, 0, .5)
    _ctx.fill(.2, .2, .9)
    _ctx.fontsize(10)
    
    for y in range(0, _ctx.HEIGHT, gridsize):
        _ctx.line(0, y, _ctx.WIDTH, y)

    for x in range(0, _ctx.HEIGHT, gridsize):
        _ctx.line(x, 0, x, _ctx.HEIGHT)

    
    for y in range(0, _ctx.HEIGHT, gridsize):
        for x in range(0, _ctx.HEIGHT, gridsize):
            _ctx.text(f"{x}, {y}", x+2, y+10, width=100)
```

I'll close this as a ticket for now, since I think scale is OK..  having better debug tools would help though, so at the very least I might see about adding this to lib somewhere.

**Comment by karstenw** [comment]
> Started looking at a little drawgrid routing you can import (and at the same time was reminded about how ugly `_ctx` looks, but I'm not sure there is a good solution for that (especially in the nodebox1/shoebot style) :)

Just reading that sentence and thinking: of course it's ugly... and it can be remedied if you ximport() the library.

IIRC all ximport does is creating the _ctx object.

When a script is run a giant dict of new keywords is made available for the script.

Theoretically the same should be possible for ximport...

Or if NB/SB redefines import

Or if a NB/SB lib just says:
`from shoebot.bowels import line,fontsize,stroke,fill`

Just thinking out loudly

I was always annoyed by the special ximport handling in Library. If a lib is a NB/SB lib it should be able to access the innards... however ugly that would get.

**Comment by stuaxo** [comment]
Yeah, I found myself wanting all of those while trying to write the grid.

Along with - if you ximport something with `draw()` it could be called automatically before your own `draw`.

`ximport` should import bot files, and it should be able to import from the current directory.

Redefining import would be nice for the user - though would need to be done carefully so as not to break things (we redefine some things, and if we did that one every import it could break stuff.


```from shoebot.bowels import line,fontsize,stroke,fill```

Would be good; I think easy in Nodebox, I'm not sure how easy it is in Shoebot.
